File: bindings/r/R/simply.R
#' Simply Fluent API for STRling
#'
#' @description
#' High-level fluent API functions for building STRling patterns without
#' manually constructing AST nodes. This provides a cleaner, more intuitive
#' interface matching the Gold Standard naming conventions across all STRling bindings.

#' Match the start of a line (^)
#'
#' @return A STRling AST node representing the start anchor
#' @export
#' @examples
#' sl_start()
sl_start <- function() {
  strling_anchor("Start")
}

#' Match the end of a line ($)
#'
#' @return A STRling AST node representing the end anchor
#' @export
#' @examples
#' sl_end()
sl_end <- function() {
  strling_anchor("End")
}

#' Match exactly n digits
#'
#' @param n Number of digits to match. Must be a positive integer.
#' @return A STRling AST node matching n digits
#' @export
#' @examples
#' sl_digit(3)  # Matches exactly 3 digits
#' sl_digit()   # Matches exactly 1 digit
sl_digit <- function(n = 1L) {
  if (!is.numeric(n) || length(n) != 1) {
    stop("sl_digit expects a single numeric value for n")
  }
  
  if (n <= 0) {
    warning(sprintf("sl_digit: n must be a positive integer, received %s. Defaulting to 1", n))
    n <- 1L
  }
  
  digit_class <- strling_character_class(
    list(strling_class_escape("d")),
    negated = FALSE
  )
  
  if (n == 1L) {
    return(digit_class)
  }
  
  strling_quantifier(digit_class, min = as.integer(n), max = as.integer(n))
}

#' Create a character class matching any of the provided characters
#'
#' @param chars A string where each character is an option to match
#' @return A STRling AST node representing a character class
#' @export
#' @examples
#' sl_any_of("-. ")  # Matches hyphen, period, or space
#' sl_any_of("abc")  # Matches a, b, or c
sl_any_of <- function(chars) {
  if (!is.character(chars) || length(chars) != 1) {
    stop(sprintf(
      "sl_any_of expects a single character string, but received: %s. Example: sl_any_of('abc')",
      paste(class(chars), collapse = ", ")
    ))
  }
  
  char_vec <- strsplit(chars, "")[[1]]
  items <- lapply(char_vec, function(ch) strling_class_literal(ch))
  
  strling_character_class(items, negated = FALSE)
}

#' Merge/concatenate multiple patterns into a sequence
#'
#' @param ... One or more STRling AST nodes to merge in sequence
#' @return A STRling AST node representing the sequence
#' @export
#' @examples
#' sl_merge(sl_start(), sl_digit(3), sl_end())
sl_merge <- function(...) {
  patterns <- list(...)
  
  if (length(patterns) == 0) {
    return(strling_sequence(list()))
  }
  
  if (length(patterns) == 1) {
    return(patterns[[1]])
  }
  
  strling_sequence(patterns)
}

#' Create a capturing group around a pattern
#'
#' @param inner The STRling AST node to wrap in a capturing group
#' @return A STRling AST node representing a capturing group
#' @export
#' @examples
#' sl_capture(sl_digit(3))
sl_capture <- function(inner) {
  strling_group(inner, capturing = TRUE)
}

#' Make a pattern optional (match 0 or 1 times)
#'
#' @param inner The STRling AST node to make optional
#' @return A STRling AST node representing an optional pattern
#' @export
#' @examples
#' sl_may(sl_any_of("-. "))  # Optional separator
sl_may <- function(inner) {
  strling_quantifier(inner, min = 0L, max = 1L)
}

#' Compile a STRling pattern to Intermediate Representation (IR)
#'
#' @param pattern A STRling AST node to compile
#' @return The compiled IR (a nested list structure)
#' @export
#' @examples
#' phone <- sl_merge(sl_start(), sl_digit(3), sl_end())
#' ir <- sl_compile(phone)
sl_compile <- function(pattern) {
  compile_ast(pattern)
}

File: bindings/php/src/Simply.php
<?php

namespace STRling;

use STRling\Core\Nodes;
use STRling\Compiler;

/**
 * STRling Simply API - Static Pattern Constructor
 * 
 * Provides a composable API for building regex patterns using static methods.
 * This API uses static methods to wrap patterns, mimicking the Python/TypeScript nesting approach.
 * 
 * Example usage:
 * ```php
 * use STRling\Simply;
 * 
 * $phone = Simply::merge(
 *     Simply::start(),
 *     Simply::capture(Simply::digit(3)),
 *     Simply::may(Simply::inChars("-. ")),
 *     Simply::capture(Simply::digit(3)),
 *     Simply::may(Simply::inChars("-. ")),
 *     Simply::capture(Simply::digit(4)),
 *     Simply::end()
 * );
 * 
 * $regex = $phone->compile();
 * ```
 * 
 * @package STRling
 */
class Simply
{
    /**
     * Create a pattern matching digits with optional repetition.
     * 
     * @param int|null $minRep Minimum number of digits (null = match exactly once)
     * @param int|null $maxRep Maximum number of digits (null = match exactly $minRep times, 0 = unlimited)
     * @return Pattern A Pattern object representing digit pattern
     */
    public static function digit(?int $minRep = null, ?int $maxRep = null): Pattern
    {
        // Create a character class containing the digit escape (\d)
        $escapeNode = new Nodes\Escape('digit');
        $charClassNode = new Nodes\CharacterClass(
            negated: false,
            members: [$escapeNode]
        );
        
        $pattern = new Pattern($charClassNode);
        return $minRep !== null ? $pattern->rep($minRep, $maxRep) : $pattern;
    }
    
    /**
     * Create a pattern matching any one character from the provided string.
     * 
     * @param string $chars The characters to match (e.g., "-. ")
     * @return Pattern A Pattern object representing a character class
     */
    public static function anyOf(string $chars): Pattern
    {
        $members = [];
        for ($i = 0; $i < strlen($chars); $i++) {
            $members[] = new Nodes\Literal($chars[$i]);
        }
        
        $charClassNode = new Nodes\CharacterClass(
            negated: false,
            members: $members
        );
        
        return new Pattern($charClassNode);
    }
    
    /**
     * Concatenate multiple patterns into a single sequential pattern.
     * 
     * @param Pattern ...$patterns One or more patterns to concatenate
     * @return Pattern A Pattern object representing the sequence
     */
    public static function merge(Pattern ...$patterns): Pattern
    {
        $nodes = array_map(fn($p) => $p->getNode(), $patterns);
        $sequenceNode = new Nodes\Sequence(parts: $nodes);
        
        return new Pattern($sequenceNode);
    }
    
    /**
     * Create a capturing group around the provided pattern.
     * 
     * @param Pattern $pattern The pattern to capture
     * @return Pattern A Pattern object representing the capture group
     */
    public static function capture(Pattern $pattern): Pattern
    {
        $groupNode = new Nodes\Group(
            capturing: true,
            body: $pattern->getNode()
        );
        
        return new Pattern($groupNode);
    }
    
    /**
     * Make the provided pattern optional (matches 0 or 1 times).
     * 
     * @param Pattern $pattern The pattern to make optional
     * @return Pattern A Pattern object representing the optional pattern
     */
    public static function may(Pattern $pattern): Pattern
    {
        $quantifierNode = new Nodes\Quantifier(
            target: $pattern->getNode(),
            min: 0,
            max: 1,
            greedy: true,
            lazy: false,
            possessive: false
        );
        
        return new Pattern($quantifierNode);
    }
    
    /**
     * Match the start of a line.
     * 
     * @return Pattern A Pattern object representing the start anchor
     */
    public static function start(): Pattern
    {
        $anchorNode = new Nodes\Anchor(at: 'Start');
        return new Pattern($anchorNode);
    }
    
    /**
     * Match the end of a line.
     * 
     * @return Pattern A Pattern object representing the end anchor
     */
    public static function end(): Pattern
    {
        $anchorNode = new Nodes\Anchor(at: 'End');
        return new Pattern($anchorNode);
    }
}

/**
 * Pattern - Wrapper class for AST nodes
 * 
 * Represents a regex pattern that can be compiled to a regex string.
 * Wraps an internal AST node and provides compilation interface.
 */
class Pattern
{
    private Nodes\Node $node;
    
    /**
     * Create a new Pattern wrapping the given AST node.
     * 
     * @param Nodes\Node $node The AST node to wrap
     */
    public function __construct(Nodes\Node $node)
    {
        $this->node = $node;
    }
    
    /**
     * Get the underlying AST node.
     * 
     * @return Nodes\Node The wrapped AST node
     */
    public function getNode(): Nodes\Node
    {
        return $this->node;
    }
    
    /**
     * Apply repetition quantifier to this pattern.
     * 
     * This method matches the Python/TypeScript API convention where 0 represents
     * unlimited repetition. For example, rep(1, 0) means "one or more".
     * 
     * @param int $minRep Minimum number of repetitions
     * @param int|null $maxRep Maximum repetitions (null = exactly $minRep, 0 = unlimited)
     * @return Pattern A new Pattern with quantifier applied
     */
    public function rep(int $minRep, ?int $maxRep = null): Pattern
    {
        // Handle unlimited repetition (0 means unlimited, matching Python/TS convention)
        if ($maxRep === 0) {
            $max = 'inf';
        } else {
            // If maxRep is null, match exactly minRep times
            $max = $maxRep ?? $minRep;
        }
        
        $quantifierNode = new Nodes\Quantifier(
            target: $this->node,
            min: $minRep,
            max: $max,
            greedy: true,
            lazy: false,
            possessive: false
        );
        
        return new Pattern($quantifierNode);
    }
    
    /**
     * Compile this pattern to an intermediate representation (IR).
     * 
     * This compiles the pattern's AST node through the compiler pipeline
     * and returns the JSON-encoded IR for demonstration purposes.
     * In production, an emitter would convert the IR to a final regex string.
     * 
     * @return string JSON-encoded IR representation
     */
    public function compile(): string
    {
        $compiler = new Compiler();
        $ir = $compiler->compile($this->node);
        return json_encode($ir, JSON_PRETTY_PRINT);
    }
}
